Per-process lock for CSSOM objects

SimonSapin opened the pull request:
<!-- Please describe your changes on the following line: -->

Before this PR, every object reflected in CSSOM is in `Arc<RwLock<_>>` to enable safe (synchronized) mutable aliasing. Acquiring all these locks has significant cost during selector matching:

* https://bugzilla.mozilla.org/show_bug.cgi?id=1311469
* https://bugzilla.mozilla.org/show_bug.cgi?id=1335941
* https://bugzilla.mozilla.org/show_bug.cgi?id=1339703

This PR introduce a mechanism to protect many objects with the same `RwLock` that only needs to be acquired once.

In Stylo, there is one such lock per process (in a `lazy_static`), used for everything.

I non-Stylo Servo, I originally intended to have one such lock per document (for author-origin stylesheets, and one per process for user-agent and user sytlesheets since they’re shared across documents, and never mutated anyway). However I failed to have the same document-specific (or pipeline-specific) `Arc` reachable from both `Document` nodes and `LayoutThread`. Recursively following callers lead me to include this `Arc` in `UnprivilegedPipelineContent`, but that needs to be serializable. So there is a second process-wide lock.

This was previously #15998, closed accidentally.

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

<!-- Either: -->
- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/16014)
<!-- Reviewable:end -->

SimonSapin committed:
Add SharedRwLock<T> and Locked<T>

SimonSapin committed:
Replace RwLock<MediaList> with shared_lock::Locked<MediaList>

SimonSapin committed:
Move shared lock acquires up the call stack. (Or is it down?)

SimonSapin committed:
More lock acquire in callers

SimonSapin committed:
Introduce a ToCssWithGuard trait

SimonSapin committed:
Replace more RwLock<MediaList> with shared_lock::Locked<MediaList>

SimonSapin committed:
Tidy

SimonSapin committed:
Replace RwLock<CssRules> with Locked<CssRules>

SimonSapin committed:
Wrap most CSS rules in Locked<_> instead of RwLock<_>

SimonSapin committed:
Replace RwLock<Keyframe> with Locked<Keyframe>

SimonSapin committed:
Replace RwLock<StyleRule> with Locked<StyleRule>

SimonSapin committed:
Move all PropertyDeclarationBlock from RwLock<_> to Locked<_>

SimonSapin commented:
@bors-servo try
r? @bholley or @emilio

bors-servo commented:
:hourglass: Trying commit c0def16580cb9ddfe8264a44b4c71d12b23b4361 with merge 913166160c02c991d3fbf810afd6ef86bfe257bd...

SimonSapin commented:
@asajeffrey, any idea how to get something not serializable both in `LayoutThead` and corresponding `Document` nodes?

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/2876)

SimonSapin commented:
A number of unexpected timeouts, possibly deadlocks with `shared_lock.read()` and `shared_lock.write()` on the same thread. The set of affected tests is mostly the same on mac and linux, so it doesn’t look *too* intermittent.

* test-css:

  <details>
  <pre>
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css-flexbox-1_dev/html/flexbox-overflow-vert-001.htm", "line": 10112, "message": null, "expected": "FAIL"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-001.htm", "line": 29958, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-002.htm", "line": 29961, "message": null, "expected": "FAIL"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-004.htm", "line": 29963, "message": null, "expected": "FAIL"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-005.htm", "line": 29968, "message": null, "expected": "FAIL"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-006.htm", "line": 29986, "message": null, "expected": "FAIL"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-008.htm", "line": 29989, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-009.htm", "line": 29993, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-012.htm", "line": 30001, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-014.htm", "line": 30014, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-015.htm", "line": 30017, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-016.htm", "line": 30021, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-019.htm", "line": 30035, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-020.htm", "line": 30036, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-021.htm", "line": 30041, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-024.htm", "line": 30043, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-025.htm", "line": 30048, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-026.htm", "line": 30070, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-028.htm", "line": 30073, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-027.htm", "line": 30074, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-029.htm", "line": 30080, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-030.htm", "line": 30098, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-033.htm", "line": 30099, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-034.htm", "line": 30103, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-035.htm", "line": 30116, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-036.htm", "line": 30126, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-037.htm", "line": 30130, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-038.htm", "line": 30132, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-039.htm", "line": 30147, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-040.htm", "line": 30148, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-041.htm", "line": 30153, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-042.htm", "line": 30154, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-043.htm", "line": 30159, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-044.htm", "line": 30182, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-045.htm", "line": 30185, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-046.htm", "line": 30186, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-047.htm", "line": 30196, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-048.htm", "line": 30210, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-050.htm", "line": 30211, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-049.htm", "line": 30212, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-051.htm", "line": 30230, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-052.htm", "line": 30251, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-053.htm", "line": 30260, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-054.htm", "line": 30277, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-055.htm", "line": 30319, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-056.htm", "line": 30320, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-059.htm", "line": 30335, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-058.htm", "line": 30336, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-057.htm", "line": 30348, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-charset-060.htm", "line": 31786, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-keywords-000.htm", "line": 31964, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-rule-005.htm", "line": 32012, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/at-rule-010.htm", "line": 32050, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/c11-import-000.htm", "line": 46773, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/c64-uri-000.htm", "line": 48366, "message": null, "expected": "FAIL"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/cascade-007.htm", "line": 48435, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/cascade-012.htm", "line": 48500, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/core-syntax-009.htm", "line": 51645, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/eof-005.htm", "line": 52974, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/eof-006.htm", "line": 52993, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/uri-001.htm", "line": 79307, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css21_dev/html4/uri-018.htm", "line": 79448, "message": null, "expected": "FAIL"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css-conditional-3_dev/html/test_group_insertRule.htm", "line": 87671, "message": null, "expected": "OK"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css-conditional-3_dev/html/001.htm", "line": 87673, "message": null, "expected": "OK"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/cssom-1_dev/html/cssimportrule.htm", "line": 88192, "message": null, "expected": "OK"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/cssom-1_dev/html/index-003.htm", "line": 88195, "message": null, "expected": "OK"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/cssom-1_dev/html/style-sheet-interfaces-001.htm", "line": 88203, "message": null, "expected": "OK"}
  </pre>
  </details>

* test-wpt:

  <details>
  <pre>
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/html/semantics/links/linktypes/alternate-css.html", "line": 68790, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/_mozilla/mozilla/stylesheet-adopt-panic.html", "line": 72956, "message": null, "expected": "PASS"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/html/semantics/document-metadata/the-style-element/style_events.html", "line": 207258, "message": null, "expected": "OK"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/old-tests/submission/Opera/script_scheduling/106-import.html", "line": 228767, "message": null, "expected": "OK"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/old-tests/submission/Opera/script_scheduling/107-import.html", "line": 228781, "message": null, "expected": "OK"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/_mozilla/css/import_serialization.html", "line": 276625, "message": null, "expected": "OK"}
  {"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/_mozilla/mozilla/out-of-order-stylesheet-loads-and-imports.html", "line": 277859, "message": null, "expected": "OK"}
  </pre>
  </details>

SimonSapin commented:
Or lot of these tests use `@import`, and I can get Servo to hang reliably with `./mach run -d -z -o /tmp/a.png 'data:text/html,<style>@import "data:,'`

SimonSapin committed:
Refactor StylesheetLoader so impls do not need to acquire a shared lock.

This fixes a deadlock:
https://github.com/servo/servo/pull/16014#issuecomment-287527067

SimonSapin commented:
New commit fixes the deadlock for my reduced test case.

@bors-servo try

bors-servo commented:
:hourglass: Trying commit e584b2e934134517d8709bd496649759bc85c5d2 with merge 8306fa5d05c4e5b8fa6c9d3cd5940f07f467739b...

SimonSapin committed:
Remove some indirection.

SimonSapin committed:
Parse stylesheets without acquiring a shared lock.

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/2918)

SimonSapin commented:
Same result on mac and linux:
```
{"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css-conditional-3_dev/html/001.htm", "line": 87760, "message": null, "expected": "OK"}
{"status": "TIMEOUT", "stack": null, "subtest": null, "action": "test_result", "test": "/css-conditional-3_dev/html/test_group_insertRule.htm", "line": 87765, "message": null, "expected": "OK"}
```

SimonSapin committed:
Fix another deadlock

SimonSapin commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 739370455fb69b59d15650e5e1df9b19e7f3a474 with merge 107a565ae3389cadb07e405d080ca8983d080678...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/6208)

SimonSapin commented:
The failure linked above is tidy. (Fixed.) Other than that, there’s one failure on mac only, that I couldn’t reproduce locally (though I’m on linux).

```
{"status": "CRASH", "stack": null, "subtest": null, "action": "test_result", "test": "/html/browsers/browsing-the-web/navigating-across-documents/javascript-url-query-fragment-components.html", "line": 131714, "message": null, "expected": "OK"}
```

Let’s see if it’s intermittent:
@bors-servo try

Also, gecko try: https://treeherder.mozilla.org/#/jobs?repo=try&revision=f07e50b0a24c0584d908711166bd283c06331c15

bors-servo commented:
:hourglass: Trying commit c01c7e69292d514f419d19736510da3deeb7a553 with merge 6f0fda351047943fc103b9b7d92d75b3382f173b...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/5847), [arm32](http://build.servo.org/builders/arm32/builds/6072), [arm64](http://build.servo.org/builders/arm64/builds/6102), [linux-dev](http://build.servo.org/builders/linux-dev/builds/6209), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/2922), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/2946), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/6211), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/6238), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/2883), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/2961), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/1375)
State: approved= try=True

jdm commented:
@SimonSapin That's #14323.

SimonSapin commented:
Ok, will do. I assume you’ve already found out in this case, but for posterity, the log was:

```
  ▶ CRASH [expected OK] /html/browsers/browsing-the-web/navigating-across-documents/javascript-url-query-fragment-components.html
  │ 
  │ VMware, Inc.
  │ Gallium 0.4 on softpipe
  │ 3.3 (Core Profile) Mesa 12.0.1
  │ called `Result::unwrap()` on an `Err` value: Error { repr: Custom(Custom { kind: UnexpectedEof, error: StringError("failed to fill whole buffer") }) } (thread RenderBackend, at src/libcore/result.rs:860)
  │ 2017-03-18 11:15:34.722 servo[3963:679522987] Metadata.framework [Error]: couldn't get the client port
  │ stack backtrace:
  │    0:        0x10dd912ce - backtrace::backtrace::trace::hef11bbe02813de13
  │    1:        0x10dd91a1c - backtrace::capture::Backtrace::new::hb5a725a088a2a2fc
  │    2:        0x10c320c50 - servo::main::_$u7b$$u7b$closure$u7d$$u7d$::hab3ba3e402ac1f0b
  │    3:        0x10ddf7459 - std::panicking::rust_panic_with_hook::hdbc3bba6a9dc0bb9
  │    4:        0x10ddf72b4 - std::panicking::begin_panic::ha249c774b9bf96e1
  │    5:        0x10ddf7222 - std::panicking::begin_panic_fmt::ha197daaf02a649ab
  │    6:        0x10ddf7187 - rust_begin_unwind
  │    7:        0x10de1f610 - core::panicking::panic_fmt::h7414cb5ce71ea361
  │    8:        0x10d98f152 - core::result::unwrap_failed::hd2d87bd00eb88923
  │    9:        0x10d9ed23f - webrender::render_backend::RenderBackend::run::h586004cebbd294f6
  │   10:        0x10d9816f8 - std::panicking::try::do_call::hfd5d9d1d1c64810e
  │   11:        0x10ddf838a - __rust_maybe_catch_panic
  │   12:        0x10d99906a - _$LT$F$u20$as$u20$alloc..boxed..FnBox$LT$A$GT$$GT$::call_box::hd4f6c63a2c1d06fd
  │   13:        0x10ddf3fe4 - std::sys::imp::thread::Thread::new::thread_start::h4008e1859fbd98b8
  │   14:     0x7fff8655c059 - _pthread_body
  │   15:     0x7fff8655bfd6 - _pthread_start
  │ ERROR:servo: called `Result::unwrap()` on an `Err` value: Error { repr: Custom(Custom { kind: UnexpectedEof, error: StringError("failed to fill whole buffer") }) }
  │ Pipeline failed in hard-fail mode.  Crashing!
  │ thread panicked while processing panic. aborting.
  └ thread panicked while processing panic. aborting.

emilio commented on components/layout_thread/lib.rs:
```diff
@@ -1057,9 +1070,17 @@ impl LayoutThread {
-                                                                                 data.stylesheets_changed);
+        let ua_stylesheets = &*UA_STYLESHEETS;
+        let ua_or_user_guard = ua_stylesheets.shared_lock.read();
+        let guards = ReadGuards {
```
Maybe rename to `StyleSheetReadGuards` or something a bit more descriptive?

SimonSapin replied:
Done.

emilio commented on components/layout_thread/lib.rs:
```diff
@@ -1333,7 +1354,14 @@ impl LayoutThread {
         };
 
-        let mut layout_context = self.build_layout_context(&*rw_data, false);
+        // Clone an Arc so that we don’t keep borrowing `self`.
```
What `Arc` is this talking about? Is this an outdated comment?

SimonSapin replied:
Ah yes, this was for per-document locks that I ended up giving up on. A removed that Arc and clone before committing but forgot to remove the comment. Fixed.

emilio commented on components/script/dom/cssmediarule.rs:
```diff
@@ -14,58 +14,71 @@ use dom::cssstylesheet::CSSStyleSheet;
+        // FIXME(SimonSapin): allow access to multiple objects with one write guard?
+        // Would need a set of usize pointer addresses or something,
+        // the same object is not accessed more than once.
+        let mqs = Arc::clone(&self.mediarule.write_with(&mut guard).media_queries);
```
Hm, this is somewhat annoying, but I guess it's not too bad.

SimonSapin replied:
We could avoid this by replacing `Arc<Locked<MediaRule>>` with `Arc<MediaRule>`. This particular lock is not useful since each of `MediaRule`’s two fields is itself `Arc<Locked<_>>`. However doing so would break some macros in `style/stylesheets.rs` and `ports/gecko/glue.rs` that generate similar code (including calls to `Locked::read_with`) for all types of rules. Maybe we should do it anyway.

emilio commented on components/script/dom/document.rs:
```diff
@@ -220,6 +222,7 @@ pub struct Document {
     scripts: MutNullableJS<HTMLCollection>,
     anchors: MutNullableJS<HTMLCollection>,
     applets: MutNullableJS<HTMLCollection>,
+    style_shared_lock: StyleSharedRwLock,
```
It'd be nice to have a description or a pointer to the rest of the documentation for this.

SimonSapin replied:
Added a doc-comment.

emilio commented on components/style/servo/mod.rs:
```diff
@@ -9,3 +9,15 @@
+    /// FIXME(SimonSapin): I wanted this to be per-document (or per-pipeline?)
+    /// but couldn’t figure out a way to get references to the same lock in both
+    /// the Document node and the LayoutThread.
+    /// Giving up to unblock.
```
I believe the layout thread can access the doc at a given point. I think it should really be per-origin, but you need to somehow propagate the state from the pipeline to script. Probably it's not too hard, but could you file and add the issue number here?

SimonSapin replied:
Filed #16027.

emilio commented on components/style/shared_lock.rs:
```diff
@@ -0,0 +1,203 @@
+    }
+
+    fn same_as(&self, other: &Self) -> bool {
+        // FIXME: Use Arc::ptr_eq once it’s stable. https://github.com/rust-lang/rust/issues/36497
```
let's use the `arc_ptr_eq` function and move this comment there.

SimonSapin replied:
Done.

emilio commented on components/style/shared_lock.rs:
```diff
@@ -0,0 +1,203 @@
+
+/// Data protect by a shared lock.
+pub struct Locked<T> {
+    shared_lock: SharedRwLock,
```
I guess it's a pain to put this lock everywhere, but I can see why it's done this way...

SimonSapin replied:
`Locked<T>` is the same size as `RwLock<T>`, in addition to `T` they store a `SharedRwLock` (a wrapper for an `Arc<_>`, which itself is a pointer) and an `AtomicUsize`, respectively.

emilio commented on components/style/shared_lock.rs:
```diff
@@ -0,0 +1,203 @@
+    trait Marker1 {}
+    impl<T: Clone> Marker1 for T {}
+    impl<'a> Marker1 for SharedRwLockReadGuard<'a> {}  // Assert SharedRwLockReadGuard: !Clone
+    impl<'a> Marker1 for SharedRwLockWriteGuard<'a> {}  // Assert SharedRwLockWriteGuard: !Clone
```
huh, clever :)

SimonSapin replied:
Credit to Niko for the original idea and Nox for showing it to me :)

emilio commented on components/style/shared_lock.rs:
```diff
@@ -0,0 +1,203 @@
+use std::ops::Deref;
+use std::sync::Arc;
+
+/// A shared read/write lock that can protect multiple objects.
```
It would be nice (here would be a decent place) to have more details here, like which locks do we have, and which ones are used for which object.

SimonSapin replied:
I originally intended this module to be a general purpose library that could be its own crate, so I kept the doc-comment generic. (Later I added `ToCssWithGuard` and `ReadGuards` there because that seemed the appropriate place for them, but they would remain in the `style` crate.)

emilio reviewed (commented):
Just a first pass with comments.

SimonSapin committed:
Rename ReadGuards to StylesheetGuards

SimonSapin commented:
Pushed fixup commits to ease review, to be autosquashed before landing.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #16029) made this pull request unmergeable. Please resolve the merge conflicts.

emilio commented:
r=me, once squashed, sorry for the lag.

And awesome work :)

SimonSapin commented:
@bors-servo r=emilio

bors-servo commented:
:pushpin: Commit d507413 has been approved by `emilio`

<!-- @bors-servo r=emilio d5074136e303544d108dc5615f96e92b854752ee -->

bors-servo commented:
:hourglass: Testing commit d5074136e303544d108dc5615f96e92b854752ee with merge bb54f0a429de0e8b8861f8071b6cf82f73622664...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/5854), [arm32](http://build.servo.org/builders/arm32/builds/6079), [arm64](http://build.servo.org/builders/arm64/builds/6109), [linux-dev](http://build.servo.org/builders/linux-dev/builds/6216), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/2929), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/2954), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/6218), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/6245), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/2892), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/2968), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/1382)
Approved by: emilio
Pushing bb54f0a429de0e8b8861f8071b6cf82f73622664 to master...
